aip/general/0129.md

```diff
@@ -0,0 +1,71 @@
+---
+id: 129
+state: draft
+created: 2023-06-21
+placement:
+  category: resource-design
+  order: 65
+---
+
+# Server-modified Values and Defaults
+
+### State-driven clients (aka Convergent clients)
+Imperative clients rely on the result of an RPC call to determine success or failure of an operation. State-driven clients (including declarative clients) are clients that use the state of an existing resource to determine actions that should be taken to bring that resource to a desired state. To determine if a set of updates was successful a state-driven client compares server side resource attributes with client defined values. The comparison feature ensures accuracy of a creation or an update but it requires services to treat the client set fields as read-only and diligently preserve those values.
+
+### Permadiff
+Services that update client-set fields communicate to state-driven clients that the desired change has not taken effect, which results in the state-driven client to continuously call the service with updates. Permadiff is an infinite loop scenario where a client continuously attempts to update a resource field or a set of fields that the API service also modifies.
+
+## Guidance
+
+### Server Side Defaults
```

Comment:
We'd need a replacement term, which would be fine if it's clear. You're defining a default as a static/transport (i.e. protobuf) default. That's reasonable- Terraform assumes the same is true, for example- but doesn't necessarily match up to what user-facing documentation treats as a default

Classifying dynamic/server-side/complex defaults as a type of default makes sense given a practical example imo. For example, a field with a description like:

> The version of Kubernetes running on this NodePool's nodes. If unspecified, it defaults as described [here](https://cloud.google.com/kubernetes-engine/versioning#specifying_node_version).

Reply:
As Riley points out there is a need for a more complex type of default that cannot be captured as a static value. That's what is meant by "dynamic default", where the value that is assigned by the server when the field is unset changes over time

Reply:
Yep, understood.

The GKE example you had seems to have a good example - the `initialVersion` which may be given values like `latest` and `1.X` . Those aren't defaults - you wouldn't even call them missing-or-empty zero values. But what is ideal is if the output_only `currentMasterVersion` contains the effective/calculated value - and the `initialValue` retained the configured `latest` or `1.X`. That's a great example of the two-field approach where the client-side and server-side aren't fighting over field ownership.

You may also go a step further and say that the default value of `initialVersion` is `latest`. That's where the string `latest` has the same meaning as the missing-or-empty zero values.

The problem has been historically that field value hasn't been recognized as a valid configuration - and the server-side has been overwriting the client's value with the result of an algorithm which may change over time. Saying the default value is `latest` is very different from saying there's a dynamic default that the service will overwrite your field with.

Re: static/transport (protobuf) default --- that's also slightly different. What protobuf calls a default is the value that's elided when marshaling and introduced when missing. That's also a bit destructive to data, because a payload can't round-trip without having default values become absent...

I'd be suggesting more that the openapi definition of default --- which is based on knowing what value is semantically equivalent to missing-or-empty zero values.

Reply:
Changed